internal/sysfs/file_windows.go

```diff
@@ -47,20 +55,49 @@ func writeFd(fd uintptr, buf []byte) (int, sys.Errno) {
 }
 
 func readSocket(h uintptr, buf []byte) (int, sys.Errno) {
+	// Poll the socket to ensure that we never perform a blocking/overlapped Read.
+	//
+	// When the socket is closed by the remote peer, wsaPoll will return n=1 and
+	// errno=0, and syscall.ReadFile will return n=0 and errno=0 -- which indicates
+	// io.EOF.
+	if n, errno := wsaPoll(
+		[]pollFd{newPollFd(h, _POLLIN, 0)}, 0); !errors.Is(errno, sys.Errno(0)) {
+		return 0, sys.UnwrapOSError(errno)
+	} else if n <= 0 {
+		return 0, sys.EAGAIN
+	}
+
+	// Properly use overlapped result.
+	//
+	// If hFile was opened with FILE_FLAG_OVERLAPPED, the following conditions are in effect:
+	//  - The lpOverlapped parameter must point to a valid and unique OVERLAPPED structure,
+	//  otherwise the function can incorrectly report that the read operation is complete.
+	//  - The lpNumberOfBytesRead parameter should be set to NULL. Use the GetOverlappedResult
+	//  function to get the actual number of bytes read. If the hFile parameter is associated
+	//  with an I/O completion port, you can also get the number of bytes read by calling the
+	//  GetQueuedCompletionStatus function.
+	//
+	// We are currently skipping checking if hFile was opened with FILE_FLAG_OVERLAPPED but using
+	// both lpOverlapped and lpNumberOfBytesRead.
+
 	var overlapped syscall.Overlapped
 	var done uint32
 	errno := syscall.ReadFile(syscall.Handle(h), buf, &done, &overlapped)
```

Comment:
Not sure if it is good to use the original input `buf` here, or better using a temporary buffer with the size up to `n` returned by `wsaPoll`. Considering it is possible for the buffer to be in use even after `syscall.ReadFile` and `getOverlappedResult` returns and by the next time caller calls this function we already lost the information in `overlapped`.

Any thoughts? @evacchi

Reply:
Actually n just returns the number of FDS that are ready, so it must be >=0. Instead of readfile i think you can just use syscall.Read and we avoid the complexity of overlapped i/o altogether because we will do a read only when there's some data ready

Reply:
`syscall.Read` very unfortunately returns an error. I think it has something to do with internally calling `syscall.ReadFile` with `overlapped` being `nil`.

Reply:
But I see what you meant there. So since `wsaPoll` does not really tell you how many bytes are ready, I am good with the current implementation.